src/main/java/seedu/duke/parsermanager/Parser.java

```diff
@@ -1,24 +1,81 @@
 package seedu.duke.parsermanager;
 
 import seedu.duke.command.Command;
-import seedu.duke.exception.ClientAlreadyPairedException;
 import seedu.duke.exception.DukeException;
-import seedu.duke.exception.EmptyDescriptionException;
 import seedu.duke.exception.EmptyDetailException;
-import seedu.duke.exception.ExistingPairException;
-import seedu.duke.exception.ExtraParametersException;
-import seedu.duke.exception.IncorrectFlagOrderException;
-import seedu.duke.exception.IncorrectListFlagsException;
-import seedu.duke.exception.InvalidBudgetFormatException;
-import seedu.duke.exception.InvalidContactNumberException;
-import seedu.duke.exception.InvalidEmailException;
-import seedu.duke.exception.InvalidIndexException;
-import seedu.duke.exception.InvalidPriceFormatException;
-import seedu.duke.exception.InvalidSingaporeAddressException;
-import seedu.duke.exception.MissingFlagException;
-import seedu.duke.exception.NoExistingPairException;
 import seedu.duke.exception.NotIntegerException;
 
+import java.util.ArrayList;
+
+import static seedu.duke.Messages.EXCEPTION;
+
 public abstract class Parser {
+    protected static final int UNIT_VALUE = 1;
+
     public abstract Command parseCommand() throws DukeException;
+
+    protected void checkForEmptyDetails(String commandDetail) throws EmptyDetailException {
+        boolean isEmptyDetail = isEmptyString(commandDetail);
+        if (isEmptyDetail) {
+            throw new EmptyDetailException(EXCEPTION);
+        }
+    }
+
+    private boolean isEmptyString(String commandDetail) {
+        return commandDetail.trim().isEmpty();
+    }
+
+    protected int[] getFlagIndexPositions(String commandDetail, String[] flags) {
+        int[] flagIndexPositions = new int[flags.length];
+
+        for (int flagIndex = 0; flagIndex < flags.length; flagIndex++) {
+            flagIndexPositions[flagIndex] = commandDetail.indexOf(flags[flagIndex]);
+        }
+        return flagIndexPositions;
+    }
+
+    protected ArrayList<String> extractCommandDetails(String rawCommandDetail, String[] flags,
+                                                    int[] flagIndexPositions) {
+        ArrayList<String> extractedCommandDetails = new ArrayList<>();
+        for (int i = 0; i < flags.length; i++) {
+            String extractedDetail;
+            if (i == flags.length - UNIT_VALUE) {
+                /* The extracted detail for the last flag starts from the char after the flag, to the end of
+                   rawCommandDetails */
+                extractedDetail = extractDetail(rawCommandDetail, flagIndexPositions[i] + flags[i].length());
```

Comment:
Perhaps can try to use a different name. The method name and the variable name seems similar and can be confusing.